File: src/src_core/fmg/core/img/AnimatedImageModel.java
package fmg.core.img;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import fmg.common.Color;
import fmg.common.geom.BoundDouble;
import fmg.common.geom.DoubleExt;
import fmg.common.geom.SizeDouble;
import fmg.common.notyfier.NotifyPropertyChanged;

/** MVC: model. Common animated image characteristics. */
public class AnimatedImageModel implements IAnimatedModel {

    public static final Color DefaultBkColor         = Color.DarkOrange();
    public static final Color DefaultForegroundColor = Color.Orchid();
    public static final int   DefaultImageSize = 100;
    public static final int   DefaultPadding = (int)(DefaultImageSize * 0.05); // 5%


    /** width and height in pixel */
    private SizeDouble _size = new SizeDouble(DefaultImageSize, DefaultImageSize);
    /** inside padding. Автоматически пропорционально регулирую при измениях размеров */
    private BoundDouble _padding = new BoundDouble(DefaultPadding);
    private Color _foregroundColor = DefaultForegroundColor;
    /** background fill color */
    private Color _backgroundColor = DefaultBkColor;
    private Color _borderColor = Color.Maroon().darker(0.5);
    private double _borderWidth = 3;
    /** 0° .. +360° */
    private double _rotateAngle;
    /** animation of polar lights */
    private boolean _polarLights = true;
    /** animation direction (example: clockwise or counterclockwise for simple rotation) */
    private boolean _animeDirection = true;
    private final AnimatedInnerModel _innerModel = new AnimatedInnerModel();
    private PropertyChangeListener innerModelListener = ev -> onInnerModelPropertyChanged(ev);
    protected NotifyPropertyChanged _notifier = new NotifyPropertyChanged(this);

    public AnimatedImageModel() {
        _innerModel.addListener(innerModelListener);
    }

    public static final String PROPERTY_PADDING          = "Padding";
    public static final String PROPERTY_BACKGROUND_COLOR = "BackgroundColor";
    public static final String PROPERTY_BORDER_COLOR     = "BorderColor";
    public static final String PROPERTY_BORDER_WIDTH     = "BorderWidth";
    public static final String PROPERTY_FOREGROUND_COLOR = "ForegroundColor";
    public static final String PROPERTY_ROTATE_ANGLE     = "RotateAngle";
    public static final String PROPERTY_POLAR_LIGHTS     = "PolarLights";
    public static final String PROPERTY_ANIME_DIRECTION  = "AnimeDirection";

    /** width and height in pixel */
    @Override
    public SizeDouble getSize() { return _size; }
    public void setSize(double widhtAndHeight) { setSize(new SizeDouble(widhtAndHeight, widhtAndHeight)); }
    @Override
    public void setSize(SizeDouble value) {
        SizeDouble old = _size;
        if (_notifier.setProperty(_size, value, PROPERTY_SIZE))
            recalcPadding(old);
    }

    /** inside padding */
    public BoundDouble getPadding() { return _padding; }
    public void setPadding(double bound) { setPadding(new BoundDouble(bound)); }
    public void setPadding(BoundDouble value) {
        if (value.getLeftAndRight() >= getSize().width)
            throw new IllegalArgumentException("Padding size is very large. Should be less than Width.");
        if (value.getTopAndBottom() >= getSize().height)
            throw new IllegalArgumentException("Padding size is very large. Should be less than Height.");
        BoundDouble paddingNew = new BoundDouble(value.left, value.top, value.right, value.bottom);
        _notifier.setProperty(_padding, paddingNew, PROPERTY_PADDING);
    }
    static BoundDouble recalcPadding(BoundDouble padding, SizeDouble current, SizeDouble old) {
        return new BoundDouble(padding.left   * current.width  / old.width,
                               padding.top    * current.height / old.height,
                               padding.right  * current.width  / old.width,
                               padding.bottom * current.height / old.height);
    }
    private void recalcPadding(SizeDouble old) {
        setPadding(recalcPadding(_padding, _size, old));
    }

    public Color getForegroundColor() { return _foregroundColor; }
    public void setForegroundColor(Color value) {
        _notifier.setProperty(_foregroundColor, value, PROPERTY_FOREGROUND_COLOR);
    }

    /** background fill color */
    public Color getBackgroundColor() { return _backgroundColor; }
    public void setBackgroundColor(Color value) {
        _notifier.setProperty(_backgroundColor, value, PROPERTY_BACKGROUND_COLOR);
    }

    public Color getBorderColor() { return _borderColor; }
    public void setBorderColor(Color value) {
        _notifier.setProperty(_borderColor, value, PROPERTY_BORDER_COLOR);
    }

    public double getBorderWidth() { return _borderWidth; }
    public void setBorderWidth(double value) {
        if (!DoubleExt.hasMinDiff(_borderWidth, value)) {
            double old = _borderWidth;
            _borderWidth = value;
            _notifier.onPropertyChanged(old, value, PROPERTY_BORDER_WIDTH);
        }
    }

    /** 0° .. +360° */
    public double getRotateAngle() { return _rotateAngle; }
    public void setRotateAngle(double value) {
        _notifier.setProperty(_rotateAngle, fixAngle(value), PROPERTY_ROTATE_ANGLE);
    }

    /** to diapason (0° .. +360°] */
    public static double fixAngle(double value) {
        return (value >= 360)
             ?              (value % 360)
             : (value < 0)
                ?           (value % 360) + 360
                :            value;
    }

    /** Image is animated? */
    @Override
    public boolean isAnimated() { return _innerModel.isAnimated(); }
    @Override
    public void setAnimated(boolean value) { _innerModel.setAnimated(value); }

    /** Overall animation period (in milliseconds) */
    @Override
    public long getAnimatePeriod() { return _innerModel.getAnimatePeriod(); }
    /** Overall animation period (in milliseconds) */
    @Override
    public void setAnimatePeriod(long value) { _innerModel.setAnimatePeriod(value); }

    /** Total frames of the animated period */
    @Override
    public int getTotalFrames() { return _innerModel.getTotalFrames(); }
    @Override
    public void setTotalFrames(int value) { _innerModel.setTotalFrames(value); }

    @Override
    public int getCurrentFrame() { return _innerModel.getCurrentFrame(); }
    @Override
    public void setCurrentFrame(int value) { _innerModel.setCurrentFrame(value); }

    public boolean isPolarLights() { return _polarLights; }
    public void setPolarLights(boolean polarLights) {
        _notifier.setProperty(_polarLights, polarLights, PROPERTY_POLAR_LIGHTS);
    }

    public boolean getAnimeDirection() { return _animeDirection; }
    public void setAnimeDirection(boolean animeDirection) {
        _notifier.setProperty(_animeDirection, animeDirection, PROPERTY_ANIME_DIRECTION);
    }

    protected void onInnerModelPropertyChanged(PropertyChangeEvent ev) {
        // refire
        _notifier.onPropertyChanged(ev.getOldValue(), ev.getNewValue(), ev.getPropertyName());
    }

    @Override
    public void close() {
        _innerModel.removeListener(innerModelListener);
        _notifier.close();
    }

    @Override
    public void addListener(PropertyChangeListener listener) {
        _notifier.addListener(listener);
    }
    @Override
    public void removeListener(PropertyChangeListener listener) {
        _notifier.removeListener(listener);
    }

}

File: src/src_core/fmg/core/img/AnimatedImageModel.cs
using System;
using System.ComponentModel;
using fmg.common;
using fmg.common.geom;
using fmg.common.notifier;

namespace fmg.core.img {

    public static class AnimatedImageModelConst {

        public static readonly Color DefaultBkColor = Color.DarkOrange; // Color.Coral; // 
        public static readonly Color DefaultForegroundColor = Color.LightSeaGreen; // Color.Orchid; // 
        public const int DefaultImageSize = 100;
        public const int DefaultPadding = (int)(DefaultImageSize * 0.05); // 5%

    }

    /// <summary> MVC: model. Common animated image characteristics. </summary>
    public abstract class AnimatedImageModel : IAnimatedModel {

        /// <summary> width and height in pixel </summary>
        private SizeDouble _size = new SizeDouble(AnimatedImageModelConst.DefaultImageSize, AnimatedImageModelConst.DefaultImageSize);
        /// <summary> inside padding </summary>
        private BoundDouble _padding = new BoundDouble(AnimatedImageModelConst.DefaultPadding);
        private Color _foregroundColor = AnimatedImageModelConst.DefaultForegroundColor;
        /// <summary> background fill color </summary>
        private Color _backgroundColor = AnimatedImageModelConst.DefaultBkColor;
        private Color _borderColor = Color.Maroon.Darker(0.5);
        private double _borderWidth = 3;
        /// <summary> 0° .. +360° </summary>
        private double _rotateAngle;

        /** animation of polar lights */
        private bool _polarLights = true;
        /** animation direction (example: clockwise or counterclockwise for simple rotation) */
        private bool _animeDirection = true;
        private readonly AnimatedInnerModel _innerModel = new AnimatedInnerModel();

        protected bool Disposed { get; private set; }
        public event PropertyChangedEventHandler PropertyChanged;
        protected readonly NotifyPropertyChanged _notifier;

        protected AnimatedImageModel() {
            _notifier = new NotifyPropertyChanged(this, ev => PropertyChanged?.Invoke(this, ev));
            _innerModel.PropertyChanged += OnInnerModelPropertyChanged;
        }

        /// <summary> width and height in pixel </summary>
        public SizeDouble Size {
            get { return _size; }
            set {
                this.CheckSize(value);
                var old = _size;
                if (_notifier.SetProperty(ref _size, value))
                    Padding = this.RecalcPadding(_padding, _size, old);
            }
        }

        /// <summary> inside padding </summary>
        public BoundDouble Padding {
            get { return _padding; }
            set {
                this.CheckPadding(value);
                _notifier.SetProperty(ref _padding, value);
            }
        }

        public Color ForegroundColor {
            get { return _foregroundColor; }
            set { _notifier.SetProperty(ref _foregroundColor, value); }
        }

        /// <summary> background fill color </summary>
        public Color BackgroundColor {
            get { return _backgroundColor; }
            set { _notifier.SetProperty(ref _backgroundColor, value); }
        }

        public Color BorderColor {
            get { return _borderColor; }
            set { _notifier.SetProperty(ref _borderColor, value); }
        }

        public double BorderWidth {
            get { return _borderWidth; }
            set {
                // _notifier.SetProperty(ref _borderWidth, value);
                if (!_borderWidth.HasMinDiff(value)) {
                    double old = _borderWidth;
                    _borderWidth = value;
                    _notifier.FirePropertyChanged<double>(old, value, nameof(this.BorderWidth));
                }
            }
        }

        /// <summary> 0° .. +360° </summary>
        public double RotateAngle {
            get { return _rotateAngle; }
            set { _notifier.SetProperty(ref _rotateAngle, FixAngle(value)); }
        }

        /// <summary> to diapason (0° .. +360°] </summary>
        internal static double FixAngle(double value) {
            return (value >= 360)
                 ? (value % 360)
                 : (value < 0)
                    ? (value % 360) + 360
                    :  value;
        }

        public bool Animated {
            get { return _innerModel.Animated; }
            set { _innerModel.Animated = value; }
        }

        //[Obsolete("oldest use")]
        //public void SetRIandRAD(int redrawInterval = 100, double rotateAngleDelta = 1.4) {
        //    double totalFrames = 360 / rotateAngleDelta;
        //    double animatePeriod = totalFrames * redrawInterval;
        //    TotalFrames = (int)totalFrames;
        //    AnimatePeriod = (long)animatePeriod;
        //}

        /// <summary> Overall animation period (in milliseconds) </summary>
        public long AnimatePeriod {
            get { return _innerModel.AnimatePeriod; }
            set { _innerModel.AnimatePeriod = value; }
        }

        /// <summary> Total frames of the animated period </summary>
        public int TotalFrames {
            get { return _innerModel.TotalFrames; }
            set { _innerModel.TotalFrames = value; }
        }

        public int CurrentFrame {
            get { return _innerModel.CurrentFrame; }
            set { _innerModel.CurrentFrame = value; }
        }

        public bool PolarLights {
            get { return _polarLights; }
            set { _notifier.SetProperty(ref _polarLights, value); }
        }

        public bool AnimeDirection {
            get { return _animeDirection; }
            set { _notifier.SetProperty(ref _animeDirection, value); }
        }

        protected void OnInnerModelPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            // refire
            _notifier.FirePropertyChanged(ev);
        }

        // <summary>  Dispose managed resources </summary>/
        protected virtual void Disposing() {
            _innerModel.PropertyChanged -= OnInnerModelPropertyChanged;
            _notifier.Dispose();
        }

        public void Dispose() {
            if (Disposed)
                return;
            Disposed = true;
            Disposing();
            GC.SuppressFinalize(this);
        }

    }

}
